### DISCUSSION TITLE
Abstracting the blender pointer for Generic Entities

### DISCUSSION THREAD

**Discussion by amrsoll** [hlorus/CAD_Sketcher#205] [Feature Requests]
This refers to most `SlvsSomeClass` classes inside `class_defines.py` but can be treated as a more generalised topic about injecting properties into classes and objects.

### Motivation

It is unclear -- appart from docstring -- what attributes the classes have, given that they get injected new attributes through the use of `slvs_entity_pointer()` right after the definition of the class.

This removes a point of contention when trying to read and understand the code.

Basically, `slvs_entity_pointer()`shortens this code

```py
class SlvsSketch(SlvsGenericEntity, PropertyGroup):

    # index pointer to the blender entity
    wp_i: IntProperty(name="wp index", default=-1)

    @property
    def wp(self):
        if self.wp_i == -1:
            logger.debug(f"No entity index for {property_name} inside element {self}")
            return
        # Get the blender enity using the index/pointer
        return bpy.context.scene.sketcher.entities.get(self.wp_i)

    @wp.setter
    def wp(self, wp_entity: SlvsWorkplane):
        # Assign a new entity based on the new entity index
        self.wp_i = wp_entity.slvs_index if wp_entity else -1
```

down to this, by modifying the class after the fact.

```py

class SlvsSketch(SlvsGenericEntity, PropertyGroup):
    [...]

slvs_entity_pointer(SlvsSketch, "wp")
```

### Description

> A clear and concise description of what you want to happen.

The SlvsGenericEntity holds an index value for itself, and has functions to get/set the blender entity based on that index/pointer. Derived classes will automatically have an index. This also makes more separation of concern and abstraction

### Alternative Solutions

> A clear and concise description of any alternative solutions or features you've considered.

A less desirable solution would be to convert class-editing into class decorators (see mockup). This highlights the modification at the top of the class instead of trailing at the end of the `class` code block.

### Mockup

(mockup) **Solution for `SlvsGenericEntity` having control of the pointer**

```py
class SlvsGenericEntity(PropertyGroup):

    index: IntProperty(name="index", default=-1)

    def entity(self):
        """The get function for the blender entity"""
        if self.index == -1:
            errmsg = f"No entity index for {property_name} inside element {self}"
            logger.info(errmsg)
            # raise IndexError(errmsg)
            return None
        else:
            return bpy.context.scene.sketcher.entities.get(index)

    def update(self, new_entity):
        """The set function for the blender entity"""
        self.index = new_entity.slvs_inex if new_entity else -1

    [...]

class SlvsWorkplane(SlvsGenericEntity):
    """Representation of a plane which is defined by an origin point
    and a normal. Workplanes are used to define the position of 2D entities
    which only store the coordinates on the plane.

    Arguments:
        p1 (SlvsPoint3D): Origin Point of the Plane
        nm (SlvsNormal3D): Normal which defines the orientation
    """
    # No longer requires to be registered

    def __init__(self, p1: SlvsPoint3D, nm: SlvsNormal3D):
        super().__init__(self)
        self.p1 = p1
        self.nm = nm

# most sublasses might still need to be registered, such as `SlvsPoint3D` 
# because they have extra properties to have registered

class SlvsSketch(SlvsGenericEntity, PropertyGroup):
    """A sketch groups 2 dimensional entities together and is used to later
    convert geometry to native blender types.

    Entities that belong to a sketch can only be edited as long as the sketch is active.

    Arguments:
        wp (SlvsWorkplane): The base workplane of the sketch
    """

    def __init__(self, wp: SlvsWorkplane):
        super().__init__(self)
        self.wp = wp

    [...]
```
This might be further simplified using [dataclasses]; Example:

```py
from dataclasses import dataclass

@dataclass
class SlvsSketch(SlvsGenericEntity, PropertyGroup):
    wp: SlvsWorkplane
    [...]
    dof: IntProperty(name="Degrees of Freedom", max=6)
    target_curve: PointerProperty(type=bpy.types.Curve)
```
However it is unclear how well it will mix with the other annotations, but perhaps `@dataclass(init=False)` might work

#### Solution with **class decorators** (not functional)

```py
@slvs_entity_pointer("p1", "Line's startpoint")
@slvs_entity_pointer("p2", "Line's endpoint")
class SlvsLine3D(SlvsGenericEntity, PropertyGroup):
    [...]
```
```py
def slvs_entity_pointer(property_name: str, doc: str="", property_kwargs: Mapping={}):
    """Class decorator which adds getter/setter/deleter functions to interact with the property through the Blender scene
    It also adds that IntProperty to the ``cls.__annotations__``
    """
    def class_wrapper(cls):
        index_prop = property_name + "_i"
        annotations = getattr(cls, "__annotations__", {}).copy()
        annotations[index_prop] = IntProperty(
            name=property_name + " index", 
            default=-1, 
            **property_kwargs
        )
        cls.__annotations__ = annotations

        def getter(self):
            index = getattr(self, index_prop)
            if index == -1:
                errmsg = f"No entity index for {property_name} inside element {self}"
                logger.info(errmsg)
                # raise IndexError(errmsg)
                return None
            else:
                return bpy.context.scene.sketcher.entities.get(index)

        def setter(self, entity):
            index = entity.slvs_index if entity else -1
            setattr(self, index_prop, index)

        def deleter(self):
            delattr(self, index_prop)
            delattr(self, property_name)

        setattr(cls, property_name, property(getter, setter, deleter, doc))
        return cls
    return class_wrapper
```

[dataclasses]: https://docs.python.org/3/library/dataclasses.html

**Comment by hlorus** [comment]
Are these functional mockups? I'm not actually sure if that code would work, i assume the SlvsWorkplane class wouldn't have a working index property when only it's parent class SlvsGenericEntity is registered.

Also i'm not sure about defining a __init__ method, this is normally used by blender internally only, no clue what arguments this method would expect.

**Reply by amrsoll** [reply]
They are not functional, (I would probably write a PR if they were), but more of a direction, to gather feedback on what you or others might think

**Reply by amrsoll** [reply]
I have tried the class decorator before, as I thought it would be relatively simple, but it did break and was really hard to debug, my dev environment isn't perfect :thinking: I don't have line breaks for example :stuck_out_tongue:

**Comment by amrsoll** [comment]
Wait, actually I just found something. Is `slvs_index` fundamentally the same indexing system as `wp_i` in the example? Isn't `wp_i == wp.slvs_index` ??

**Reply by hlorus** [reply]
Well yes, the entity pointers work by simply storing the "slvs_index" of an entity.

**Reply by hlorus** [reply]
Slvs_entity_pointer simply adds a getter and setter which looks up/ stores the entity by its index.

**Reply by amrsoll** [reply]
and just to be clear, these pointers are actually related to blender, and have nothing to do with solvesys? Solvesys doesn't use those indexes?

**Reply by amrsoll** [reply]
I am starting to get a better picture, but why did you chose to have `wp_i` as an attribute instead of `wp` ? Is it because of the `__init__` method?

**Reply by amrsoll** [reply]
*as an attribute to `SlvsSketch`

**Reply by hlorus** [reply]
> and just to be clear, these pointers are actually related to blender, and have nothing to do with solvesys? Solvesys doesn't use those indexes?

Yes they aren't directly used by the solver. However there are a few things that get stored in a global list based on the indices. Afaik also the solver handle. This isn't strictly needed as the solver isn't persistent, this was mostly an attempt to optimize things by not having to recreate the whole solversystem everytime...

**Reply by hlorus** [reply]
> I am starting to get a better picture, but why did you chose to have `wp_i` as an attribute instead of `wp` ? Is it because of the `__init__` method?

The "*_i" names are basically implicitly created IntPointers which usually aren't access directly. There's always a getter/setter function for convinient access.

**Comment by amrsoll** [comment]
After digging and debugging, I am starting to understand the route taken to define those attributes. Yes it does look like the `PropertyGroup` get destroyed as soon as the execution unit is over. Therefore you would have to re-assign the attributes every time, or regenerate them from the Blender data `bpy.data.x` or from the provided `context`, which is what the `@property` functions are doing.

No `__init__()` calls will be made with extra arguments, so we can't assign them that way either.

Blender API docs: [Integration through classes]

> Notice these classes don’t define an `__init__(self)` function. While `__init__()` and `__del__()` will be called if defined, the class instances lifetime only spans the execution. So a panel for example will have a new instance for every redraw, for this reason there is rarely a cause to store variables in the panel instance. Instead, persistent variables should be stored in Blender’s data so that the state can be restored when Blender is restarted.

Because we basically **don't have persistence** (everything has to be regenerated from primitive datatypes in blender data), I think a **functional programming approach is more desirable** than Object Oriented (OOP)

Or I could try to make this heavy pointer gymnastics work, perhaps with a `.from_pointer(pointer: int)` function that can create the instance from existing data.

[Integration through classes]: https://docs.blender.org/api/current/info_overview.html#integration-through-classes

**Comment by amrsoll** [comment]
So after a lot more research, it seems that [`PointerProperty`] is exactly what is needed to abstract the slvs indexes. I will submit a PR with a major refactor that includes switching to `PointerProperty`.

Example :

```py
Point3DProperty = partial(PointerProperty, type=SlvsPoint3D)
"""Returns a new SlvsPoint3D property definition."""


class SlvsLine3D(shape.Line, SlvsGenericEntity, PropertyGroup, GeometryClassifier):
    """Representation of a line in 3D Space.

    Arguments:
        p1 (SlvsPoint3D): Line's startpoint
        p2 (SlvsPoint3D): Line's endpoint
    """

    p1: Point3DProperty(name="p1", description="Start point")
    p2: Point3DProperty(name="p2", description="End point")

    [...]
```

[`PointerProperty`]: https://docs.blender.org/api/current/bpy.props.html#bpy.props.PointerProperty

**Reply by hlorus** [reply]
That would be ideal IMO but as far as i know PointerProperties cannot be used to point to an element of a CollectionProperty...

**Reply by amrsoll** [reply]
Yeah I thought I had resolved most stuff so I don't have to link pre-existing items, but that is the case for workplanes, and I haven't found a way around that yet

**Reply by amrsoll** [reply]
Actually yeah you are totally right, It would be sweet to initialise PointerProperties to reference a pre-existing item, which would also allow to point to the same item from different variables (for example when sharing a point between entities)

It really seems shortsighted from the blender API which makes it treacherous to develop add-ons :roll_eyes: maybe I can find a very hacky way to have a proper reassignable PointerProperty.

**Reply by amrsoll** [reply]
Quick untested research would show we can have a reassignable `PointerProperty` if we use subclasses of [`bpy.types.ID`](https://docs.blender.org/api/current/bpy.types.ID.html#bpy.types.ID) ; for example [`bpy.types.Curve`](https://docs.blender.org/api/current/bpy.types.Curve.html#bpy.types.Curve).

```py
my_prop = PointerProperty(type=bpy.types.Curve)
```

However, using native Blender objects would require to refactor a lot more than simply switching to `PointerProperty`; but it could come with a lot of niceties (pre-coded functionalities)

Note: `bpy.types.Curve` can be either 2D or 3D, see [`bpy.types.Curve.dimensions`](https://docs.blender.org/api/current/bpy.types.Curve.html#bpy.types.Curve.dimensions)

**Reply by hlorus** [reply]
Could be interesting to base the addon on an existing geometry type. This could potentially remove the need of having to implement drawing and selection of entities. I've considered this a few times already, i think it isn't trivial however with the new curve data structure it could be worth to give this another look.
